File: hindsight-clients/typescript/tests/images.test.ts
import { HindsightClient, sdk } from "../src";

const ok = (data: unknown, headers?: Headers) => ({
  data,
  error: undefined,
  response: new Response(null, { status: 200, headers }),
});

afterEach(() => jest.restoreAllMocks());

test("ordered image multipart metadata, recall map, management and idempotency", async () => {
  const retain = jest
    .spyOn(sdk, "imageRetain")
    .mockResolvedValue(ok({ operation_id: "op" }) as never);
  const recall = jest
    .spyOn(sdk, "recallMemories")
    .mockResolvedValue(ok({ results: [], image_assets: { doc: [{ asset_id: "a/1" }] } }) as never);
  jest.spyOn(sdk, "listImageAssets").mockResolvedValue(ok({ items: [], total: 0 }) as never);
  jest.spyOn(sdk, "deleteImageAsset").mockResolvedValue({
    data: undefined,
    error: undefined,
    response: new Response(null, { status: 204 }),
  } as never);
  jest.spyOn(sdk, "getImageAsset").mockResolvedValue(
    ok(
      new Blob(["bytes"], { type: "image/jpeg" }),
      new Headers({
        "content-type": "image/jpeg",
        "content-length": "5",
        "x-hindsight-image-sha256": "a".repeat(64),
        "x-hindsight-image-width": "8",
        "x-hindsight-image-height": "6",
        "x-hindsight-asset-created-at": "2026-01-01T00:00:00Z",
        "x-hindsight-asset-updated-at": "2026-01-01T00:00:00Z",
      })
    ) as never
  );
  const client = new HindsightClient({ baseUrl: "http://unused" });
  const first = new Blob(["first"], { type: "image/jpeg" });
  const second = new Blob(["second"], { type: "image/png" });
  await client.retainImages(
    "bank",
    [
      { file: first, assetId: "a/1", context: "first" },
      { file: second, assetId: "b/2", context: "second" },
    ],
    { idempotencyKey: "idem" }
  );
  const request = retain.mock.calls[0][0] as any;
  expect(request.body.files).toEqual([first, second]);
  expect(JSON.parse(request.body.request).images).toEqual([
    { asset_id: "a/1", context: "first" },
    { asset_id: "b/2", context: "second" },
  ]);
  expect(request.headers["Idempotency-Key"]).toBe("idem");

  const recalled = await client.recall("bank", "query", { includeImageAssets: true });
  expect((recall.mock.calls[0][0] as any).body.include.image_assets).toBe(true);
  expect(recalled.image_assets?.doc[0].asset_id).toBe("a/1");
  await client.listImageAssets("bank", { documentId: "doc" });
  const downloaded = await client.getImageAsset("bank", "a/1");
  expect(downloaded.descriptor.document_ids).toEqual([]);
  await client.deleteImageAsset("bank", "a/1");
});

test("transfer returns a stream and old services fail explicitly", async () => {
  const client = new HindsightClient({ baseUrl: "http://unused" });
  const stream = new ReadableStream<Uint8Array>();
  (client as any).client.get = jest.fn().mockResolvedValue({
    data: stream,
    error: undefined,
    response: { ok: true, status: 200, body: stream },
  });
  expect(await client.exportDocumentsStream("bank")).toBe(stream);

  jest.spyOn(sdk, "listImageAssets").mockResolvedValue({
    data: undefined,
    error: { detail: "not supported" },
    response: new Response(null, { status: 404 }),
  } as never);
  await expect(client.listImageAssets("bank")).rejects.toEqual(
    expect.objectContaining({ statusCode: 404 })
  );
});

test("image download rejects incomplete descriptor headers", async () => {
  jest.spyOn(sdk, "getImageAsset").mockResolvedValue(
    ok(
      new Blob(["bytes"], { type: "image/jpeg" }),
      new Headers({
        "content-type": "image/jpeg",
        "content-length": "5",
      })
    ) as never
  );
  const client = new HindsightClient({ baseUrl: "http://unused" });
  await expect(client.getImageAsset("bank", "asset")).rejects.toThrow(
    "x-hindsight-asset-created-at"
  );
});

File: hindsight-clients/go/images_test.go
package hindsight

import (
	"bytes"
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"os"
	"strings"
	"testing"
)

func testClient(server *httptest.Server) *APIClient {
	cfg := NewConfiguration()
	cfg.Servers = ServerConfigurations{{URL: server.URL}}
	return NewAPIClient(cfg)
}

func TestImageMultipartOrderHeadersAndManagement(t *testing.T) {
	var parts []string
	var requestJSON, idempotency string
	var imagePath string
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch r.Method {
		case http.MethodPost:
			idempotency = r.Header.Get("Idempotency-Key")
			reader, err := r.MultipartReader()
			if err != nil {
				t.Fatal(err)
			}
			for {
				part, err := reader.NextPart()
				if err == io.EOF {
					break
				}
				data, _ := io.ReadAll(part)
				if part.FormName() == "files" {
					parts = append(parts, string(data))
				} else if part.FormName() == "request" {
					requestJSON = string(data)
				}
			}
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusAccepted)
			_, _ = io.WriteString(w, `{"operation_id":"op","document_id":"doc","image_assets":[]}`)
		case http.MethodGet:
			if strings.HasSuffix(r.URL.Path, "/image-assets") {
				w.Header().Set("Content-Type", "application/json")
				_, _ = io.WriteString(w, `{"items":[],"total":0,"limit":100,"offset":0}`)
				return
			}
			imagePath = r.URL.EscapedPath()
			w.Header().Set("Content-Type", "image/jpeg")
			w.Header().Set("X-Hindsight-Image-SHA256", strings.Repeat("a", 64))
			w.Header().Set("X-Hindsight-Image-Width", "8")
			w.Header().Set("X-Hindsight-Image-Height", "6")
			w.Header().Set("X-Hindsight-Asset-Created-At", "2026-01-01T00:00:00Z")
			w.Header().Set("X-Hindsight-Asset-Updated-At", "2026-01-01T00:00:00Z")
			_, _ = io.WriteString(w, "bytes")
		case http.MethodDelete:
			w.WriteHeader(http.StatusNoContent)
		}
	}))
	defer server.Close()

	first, _ := os.CreateTemp(t.TempDir(), "first-*.jpg")
	second, _ := os.CreateTemp(t.TempDir(), "second-*.png")
	_, _ = first.WriteString("first")
	_, _ = second.WriteString("second")
	_, _ = first.Seek(0, io.SeekStart)
	_, _ = second.Seek(0, io.SeekStart)
	request := `{"images":[{"asset_id":"a/1"},{"asset_id":"b/2"}]}`
	client := testClient(server)
	_, _, err := client.ImagesAPI.ImageRetain(context.Background(), "bank").
		Files([]*os.File{first, second}).Request(request).IdempotencyKey("idem").Execute()
	if err != nil {
		t.Fatal(err)
	}
	if strings.Join(parts, ",") != "first,second" || requestJSON != request || idempotency != "idem" {
		t.Fatalf("multipart contract mismatch: parts=%v request=%s idempotency=%s", parts, requestJSON, idempotency)
	}
	listed, _, err := client.ImagesAPI.ListImageAssets(context.Background(), "bank").DocumentId("doc/one").Execute()
	if err != nil || listed.Total != 0 {
		t.Fatalf("list failed: listed=%#v err=%v", listed, err)
	}
	downloaded, err := client.GetManagedImageAsset(context.Background(), "bank", "folder/a.jpg")
	if err != nil {
		t.Fatal(err)
	}
	if string(downloaded.Content) != "bytes" || len(downloaded.Descriptor.DocumentIds) != 0 || !strings.Contains(imagePath, "folder%2Fa.jpg") {
		t.Fatalf("download contract mismatch: %#v", downloaded)
	}
	if _, err := client.ImagesAPI.DeleteImageAsset(context.Background(), "bank", "a/1").Execute(); err != nil {
		t.Fatal(err)
	}
	var recalled RecallResponse
	if err := json.Unmarshal([]byte(`{"results":[],"image_assets":{"doc":[]}}`), &recalled); err != nil {
		t.Fatal(err)
	}
	if recalled.ImageAssets["doc"] == nil {
		t.Fatal("recall image map was not decoded")
	}
}

func TestDocumentTransferHelpersStreamReadersAndWriters(t *testing.T) {
	archive := bytes.Repeat([]byte("archive"), 1024)
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method == http.MethodGet {
			_, _ = w.Write(archive)
			return
		}
		if err := r.ParseMultipartForm(1 << 20); err != nil {
			t.Fatal(err)
		}
		file, _, err := r.FormFile("file")
		if err != nil {
			t.Fatal(err)
		}
		uploaded, _ := io.ReadAll(file)
		if !bytes.Equal(uploaded, archive) {
			t.Fatal("uploaded archive changed")
		}
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusAccepted)
		_ = json.NewEncoder(w).Encode(DocumentImportSubmitResponse{OperationId: "op"})
	}))
	defer server.Close()
	client := testClient(server)
	var output bytes.Buffer
	if err := client.ExportDocumentsTo(context.Background(), "bank", &output); err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(output.Bytes(), archive) {
		t.Fatal("downloaded archive changed")
	}
	result, err := client.ImportDocumentsFrom(context.Background(), "bank", "transfer.zip", bytes.NewReader(archive), "skip")
	if err != nil || result.OperationId != "op" {
		t.Fatalf("import failed: result=%#v err=%v", result, err)
	}
}

type failOnRead struct {
	read bool
}

func (reader *failOnRead) Read(_ []byte) (int, error) {
	reader.read = true
	return 0, io.EOF
}

func TestImportDocumentsDoesNotStartSourceReadWhenRequestIsInvalid(t *testing.T) {
	cfg := NewConfiguration()
	cfg.Servers = ServerConfigurations{{URL: ":"}}
	client := NewAPIClient(cfg)
	source := &failOnRead{}

	if _, err := client.ImportDocumentsFrom(
		context.Background(), "bank", "archive.zip", source, "skip",
	); err == nil {
		t.Fatal("expected invalid request URL to fail")
	}
	if source.read {
		t.Fatal("source was read before the request was constructed")
	}
}
